### PR TITLE
Generalise associative operator parsing

### PR DESCRIPTION
**Pull request by nagisa**
This commit generalises parsing of associative operators from left-associative
only (with some ugly hacks to support right-associative assignment) to properly
left/right-associative operators.

Parsing is still is not general enough to handle non-associative,
non-highest-precedence prefix or non-highest-precedence
postfix operators (e.g. `..` range syntax) and should be made to be.

Lastly, this commit adds support for parsing right-associative `<-` (left arrow)
operator with precedence higher than assignment as the operator for placement-in
feature.

---

This PR still needs various non-parser changes (e.g. src/grammar and tests) and I’m still working on these; the meat of the PR can already be reviewed, though, I think.

Please review carefully. I made sure that quirks I have discovered so far are preserved (see e.g. https://github.com/rust-lang/rust/issues/29071) and am looking for more corner cases as I continue to work on tests et al, but there may be something I haven’t noticed or accounted for.

EDIT: I’m also not sure I managed to preserve all the semantics with the range operator inside non-trivial expressions since these are a mess at the moment. Crater runs would be nice.

### CODE REVIEW

**Comment by nagisa** [comment]
Something I found is that `x = 2..3 = y` now parses as `(x = 2..3) = y` while it used to fail before. The expression still fails to compile later with `E0070` anyway, though.

Something that doesn’t parse anymore and is a bug is `x = ..y` and `x = ..`. I’ll tackle these soon.

**Comment by nagisa** [comment]
Something that also needs to be fixed:

```
test.rs:45:5: 45:13 warning: unnecessary parentheses around assigned value, #[warn(unused_parens)] on by default
test.rs:45     (x <- y) <- z;
               ^~~~~~~~
```

**Comment by alexcrichton** [comment]
Running a crater build to evaluate these changes

**Comment by pnkfelix** [comment]
> Something that doesn’t parse anymore and is a bug is x = ..y and x = ... I’ll tackle these soon.

@nagisa So this is now fixed with your latest commit ?

**Comment by nagisa** [comment]
Į believe it is.

**Comment by alexcrichton** [comment]
Crater reports [two regressions](https://gist.github.com/alexcrichton/343c7b21811126567457) one of which I believe is spurious and one of which I believe is legitimate.

**Comment by nagisa** [comment]
Thanks, @alexcrichton!

**Comment by nagisa** [comment]
This PR is now in a pretty good spot, I think. The underlying issue behind the regression was just me forgetting to move restriction filtering over from the old code. This also helped to fix issues I had with semi-statement things I hacked around previously.

**Comment by pnkfelix** [comment]
(i won't get to this review until monday; hopefully that's okay.)

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -2703,86 +2779,24 @@ impl<'a> Parser<'a> {
...
-        }
+    /// Parse prefix-forms of range notation: `..expr` and `..`
+    fn parse_prefix_range_expr(&mut self) -> PResult<P<Expr>> {
+        debug_assert!(self.token == token::DotDot);
+        let lo = self.span.lo;
+        let mut hi = self.span.hi;
+        try!(self.bump());
+        let opt_end = if self.is_at_start_of_range_notation_rhs() {
+            // RHS must be parsed with more associativity than DotDot.
+            let next_prec = AssocOp::from_token(&token::DotDot).unwrap().precedence() + 1;
```
Comment (nagisa) [inline src/libsyntax/parse/parser.rs]: NB: this changes the previous bahaviour of `..` from 

> This has the same precedence as assignment expressions
> (much lower than other prefix expressions) to be consistent
> // with the postfix-form 'expr..'”

To having more precedence than `<-` and `=` but still less than any other binary operator.

**Comment by nagisa** [comment]
NB: this still needs changes to `src/grammar` and I had planned to add more tests for quirky cases such as `x..y=z` not parsing (it did not parse before, and AFAIR it does not parse after changes).

Got distracted by university things, sadly.

**Comment by nagisa** [comment]
I pushed `src/grammar` change, but honestly have no idea what I’m doing since I’m just following grammar for similarly behaving equalities.

Going from master to after the change, make check-grammar reports 2 less mismatches (from 132 down to 130).

**Comment by pnkfelix** [comment]
> EDIT: ah also, the in PLACE {EXPR} warning is removed in nagisa@58c299f

Oh, okay, sorry.  Sometimes the commit-by-commit reviewing strategy works well, but I need to learn to look ahead in the queue for things like this.

**Comment by nagisa** [comment]
> Oh, okay, sorry. Sometimes the commit-by-commit reviewing strategy works well, but I need to learn to look ahead in the queue for things like this.

I removed it after your comment, just made a note that I fixed it.

**Comment by pnkfelix** [comment]
lgtm

**Pull request merged**